### PR TITLE
fix disk_list

### PR DESCRIPTION
**Pull request by simonfuhrer**
Any subsequent changes to a virtual machine will cause the disks to be deleted. 

This PR fixes this problem.

### CODE REVIEW

**Comment by marinsalinas** [comment]
@simonfuhrer do you have any configuration to see how this change works?

**Comment by simonfuhrer** [comment]
@marinsalinas this change does not require any special configuration. It works as described in the docs 

it no longer deletes the disks if the name of the VM should be changed for example.

In the current release every change (or better every update request) causes the disks to be deleted.

**Comment by marinsalinas** [comment]
@simonfuhrer I'm testing your work and the changes broke some tests.

First, some attributes need the VM turned off that's why we turn off the VM before updating and then turn on after that.

Also some acceptance test are failing after your changes.

Here is the log:

```hcl
testing.go:518: Step 0 error: After applying this step, the plan was not empty:

        DIFF:

        UPDATE: nutanix_virtual_machine.vm1
          disk_list.0.data_source_reference.#:      "0" => "1"
          disk_list.0.data_source_reference.0.kind: "" => "image"
          disk_list.0.data_source_reference.0.uuid: "" => "e3d1cfea-55e2-424b-8bb5-46bac816cc36"

        STATE:

        data.nutanix_clusters.clusters:
          ID = 5d66621a-11d1-43c1-8fe6-5a11e718dac1
          provider = provider.nutanix
          api_version = 3.1
          entities.# = 2
          entities.0.analysis_vm_efficiency_map.% = 5
          entities.0.analysis_vm_efficiency_map.bully_vm_num = 0
          entities.0.analysis_vm_efficiency_map.constrained_vm_num = 0
          entities.0.analysis_vm_efficiency_map.dead_vm_num = 9
          entities.0.analysis_vm_efficiency_map.inefficient_vm_num = 11
          entities.0.analysis_vm_efficiency_map.overprovisioned_vm_num = 2
          entities.0.api_version =
          entities.0.authorized_public_key_list.# = 0
          entities.0.build.% = 6
          entities.0.build.build_type = release
          entities.0.build.commit_date =
          entities.0.build.commit_id = 07dbff43ac5a9f4edcd0c7ce1dccbc25a1c098b5
          entities.0.build.full_version = el7.3-release-euphrates-5.6-stable-07dbff43ac5a9f4edcd0c7ce1dccbc25a1c098b5
          entities.0.build.short_commit_id = 07dbff
          entities.0.build.version = 5.6
          entities.0.ca_certificate_list.# = 0
          entities.0.categories.% = 0
          entities.0.certification_signing_info.% = 0
          entities.0.client_auth.% = 0
          entities.0.cluster_arch = X86_64
          entities.0.domain_server_credentials.% = 0
          entities.0.domain_server_name =
          entities.0.domain_server_nameserver =
          entities.0.enabled_feature_list.# = 3
          entities.0.enabled_feature_list.0 = PASSWORD_REMOTE_LOGIN_ENABLED
          entities.0.enabled_feature_list.1 = SHADOW_CLONES_ENABLED
          entities.0.enabled_feature_list.2 = NETWORK_FLOW_STATS_COLLECTOR_ENABLED
          entities.0.encryption_status = NOT_SUPPORTED
          entities.0.external_data_services_ip =
          entities.0.external_ip =
          entities.0.external_subnet = 172.0.0.0/255.0.0.0
          entities.0.gpu_driver_version =
          entities.0.http_proxy_list.# = 0
          entities.0.http_proxy_whitelist.# = 0
          entities.0.internal_subnet = 172.0.0.0/255.0.0.0
          entities.0.is_available = true
          entities.0.management_server_list.# = 0
          entities.0.masquerading_ip =
          entities.0.masquerading_port = 0
          entities.0.metadata.% = 7
          entities.0.metadata.creation_time = 0001-01-01 00:00:00 +0000 UTC
          entities.0.metadata.kind = cluster
          entities.0.metadata.last_update_time = 0001-01-01 00:00:00 +0000 UTC
          entities.0.metadata.name =
          entities.0.metadata.spec_hash =
          entities.0.metadata.spec_version = 0
          entities.0.metadata.uuid = 27db5fb2-2aa0-48c4-bd2a-2fccdb8fad36
          entities.0.name = test-jon-gcp
          entities.0.name_server_ip_list.# = 1
          entities.0.name_server_ip_list.0 = 8.8.8.8
          entities.0.nfs_subnet_whitelist.# = 0
          entities.0.nodes.# = 2
          entities.0.nodes.0.ip = 127.0.0.1
          entities.0.nodes.0.type = AHV
          entities.0.nodes.0.version =
          entities.0.nodes.1.ip = 172.16.1.4
          entities.0.nodes.1.type = AHV
          entities.0.nodes.1.version =
          entities.0.ntp_server_ip_list.# = 1
          entities.0.ntp_server_ip_list.0 = 8.8.8.8
          entities.0.operation_mode = NORMAL
          entities.0.owner_reference.% = 0
          entities.0.project_reference.% = 0
          entities.0.service_list.# = 1
          entities.0.service_list.0 = AOS
          entities.0.smtp_server_address.% = 0
          entities.0.smtp_server_credentials.% = 0
          entities.0.smtp_server_email_address =
          entities.0.smtp_server_proxy_type_list.# = 0
          entities.0.smtp_server_type =
          entities.0.software_map_ncc.% = 3
          entities.0.software_map_ncc.software_type = NCC
          entities.0.software_map_ncc.status = INSTALLED
          entities.0.software_map_ncc.version = ncc-3.5.1
          entities.0.software_map_nos.% = 3
          entities.0.software_map_nos.software_type = NOS
          entities.0.software_map_nos.status = INSTALLED
          entities.0.software_map_nos.version = 5.6
          entities.0.ssl_key_expire_datetime =
          entities.0.ssl_key_name =
          entities.0.ssl_key_signing_info.% = 0
          entities.0.ssl_key_type =
          entities.0.state = COMPLETE
          entities.0.supported_information_verbosity = BASIC_PLUS_CORE_DUMP
          entities.0.timezone = UTC
          entities.1.analysis_vm_efficiency_map.% = 0
          entities.1.api_version =
          entities.1.authorized_public_key_list.# = 0
          entities.1.build.% = 6
          entities.1.build.build_type = release
          entities.1.build.commit_date = 2018-08-02T00:29:56Z
          entities.1.build.commit_id = aedfa93d5ad7135f43097c0a3e3010d8f7c5e8ca
          entities.1.build.full_version = el7.3-release-euphrates-5.8.1-stable-aedfa93d5ad7135f43097c0a3e3010d8f7c5e8ca
          entities.1.build.short_commit_id = aedfa9
          entities.1.build.version = 5.8.1
          entities.1.ca_certificate_list.# = 0
          entities.1.categories.% = 0
          entities.1.certification_signing_info.% = 0
          entities.1.client_auth.% = 0
          entities.1.cluster_arch = X86_64
          entities.1.domain_server_credentials.% = 0
          entities.1.domain_server_name =
          entities.1.domain_server_nameserver =
          entities.1.enabled_feature_list.# = 2
          entities.1.enabled_feature_list.0 = PASSWORD_REMOTE_LOGIN_ENABLED
          entities.1.enabled_feature_list.1 = SHADOW_CLONES_ENABLED
          entities.1.encryption_status = NOT_SUPPORTED
          entities.1.external_data_services_ip =
          entities.1.external_ip =
          entities.1.external_subnet = 172.0.0.0/255.0.0.0
          entities.1.gpu_driver_version =
          entities.1.http_proxy_list.# = 0
          entities.1.http_proxy_whitelist.# = 0
          entities.1.internal_subnet = 172.0.0.0/255.0.0.0
          entities.1.is_available = true
          entities.1.management_server_list.# = 0
          entities.1.masquerading_ip =
          entities.1.masquerading_port = 0
          entities.1.metadata.% = 7
          entities.1.metadata.creation_time = 0001-01-01 00:00:00 +0000 UTC
          entities.1.metadata.kind = cluster
          entities.1.metadata.last_update_time = 0001-01-01 00:00:00 +0000 UTC
          entities.1.metadata.name =
          entities.1.metadata.spec_hash =
          entities.1.metadata.spec_version = 0
          entities.1.metadata.uuid = 706ff6fb-0415-4830-958e-c92302297420
          entities.1.name = Unnamed
          entities.1.name_server_ip_list.# = 1
          entities.1.name_server_ip_list.0 = 8.8.8.8
          entities.1.nfs_subnet_whitelist.# = 0
          entities.1.nodes.# = 0
          entities.1.ntp_server_ip_list.# = 1
          entities.1.ntp_server_ip_list.0 = 8.8.8.8
          entities.1.operation_mode = NORMAL
          entities.1.owner_reference.% = 0
          entities.1.project_reference.% = 0
          entities.1.service_list.# = 1
          entities.1.service_list.0 = PRISM_CENTRAL
          entities.1.smtp_server_address.% = 0
          entities.1.smtp_server_credentials.% = 0
          entities.1.smtp_server_email_address =
          entities.1.smtp_server_proxy_type_list.# = 0
          entities.1.smtp_server_type =
          entities.1.software_map_ncc.% = 3
          entities.1.software_map_ncc.software_type = NCC
          entities.1.software_map_ncc.status = INSTALLED
          entities.1.software_map_ncc.version = ncc-3.5.3.1
          entities.1.software_map_nos.% = 3
          entities.1.software_map_nos.software_type = NOS
          entities.1.software_map_nos.status = INSTALLED
          entities.1.software_map_nos.version = 5.8.1
          entities.1.ssl_key_expire_datetime =
          entities.1.ssl_key_name =
          entities.1.ssl_key_signing_info.% = 0
          entities.1.ssl_key_type =
          entities.1.state = COMPLETE
          entities.1.supported_information_verbosity =
          entities.1.timezone = UTC
        nutanix_image.cirros-034-disk:
          ID = e3d1cfea-55e2-424b-8bb5-46bac816cc36
          provider = provider.nutanix
          api_version = 3.1
          availability_zone_reference.% = 0
          categories.% = 0
          checksum.% = 0
          cluster_reference.% = 0
          description = heres a tiny linux image, not an iso, but a real disk!
          image_type = DISK_IMAGE
          metadata.% = 7
          metadata.creation_time = 2018-10-17 20:08:40 +0000 UTC
          metadata.kind = image
          metadata.last_update_time = 2018-10-17 20:08:48 +0000 UTC
          metadata.name =
          metadata.spec_hash =
          metadata.spec_version = 1
          metadata.uuid = e3d1cfea-55e2-424b-8bb5-46bac816cc36
          name = test-image-dou-7532894564480512240
          owner_reference.% = 3
          owner_reference.kind = user
          owner_reference.name = admin
          owner_reference.uuid = 00000000-0000-0000-0000-000000000000
          retrieval_uri_list.# = 1
          retrieval_uri_list.0 = https://127.0.0.1:9440/api/nutanix/v3/images/e3d1cfea-55e2-424b-8bb5-46bac816cc36/file
          size_bytes = 46137344
          source_uri = http://download.cirros-cloud.net/0.4.0/cirros-0.4.0-x86_64-disk.img
          state = COMPLETE
          version.% = 0
        nutanix_virtual_machine.vm1:
          ID = dbc5818c-2e82-44fc-bfd7-2691ed68624d
          provider = provider.nutanix
          api_version = 3.1
          availability_zone_reference.% = 0
          boot_device_disk_address.% = 0
          boot_device_mac_address =
          boot_device_order_list.# = 0
          categories.% = 0
          cluster_reference.% = 2
          cluster_reference.kind = cluster
          cluster_reference.uuid = 27db5fb2-2aa0-48c4-bd2a-2fccdb8fad36
          description =
          disk_list.# = 4
          disk_list.0.data_source_reference.# = 0
          disk_list.0.device_properties.# = 1
          disk_list.0.device_properties.0.device_type = DISK
          disk_list.0.device_properties.0.disk_address.# = 1
          disk_list.0.device_properties.0.disk_address.0.adapter_type = SCSI
          disk_list.0.device_properties.0.disk_address.0.device_index = 0
          disk_list.0.disk_size_bytes = 46137344
          disk_list.0.disk_size_mib = 44
          disk_list.0.uuid = 68eba51c-24ef-4681-aaf9-c306a671cd8b
          disk_list.0.volume_group_reference.# = 0
          disk_list.1.data_source_reference.# = 0
          disk_list.1.device_properties.# = 1
          disk_list.1.device_properties.0.device_type = DISK
          disk_list.1.device_properties.0.disk_address.# = 1
          disk_list.1.device_properties.0.disk_address.0.adapter_type = SCSI
          disk_list.1.device_properties.0.disk_address.0.device_index = 1
          disk_list.1.disk_size_bytes = 104857600
          disk_list.1.disk_size_mib = 100
          disk_list.1.uuid = 2cb6c560-e994-4250-bbd3-a7579fe56b8b
          disk_list.1.volume_group_reference.# = 0
          disk_list.2.data_source_reference.# = 0
          disk_list.2.device_properties.# = 1
          disk_list.2.device_properties.0.device_type = DISK
          disk_list.2.device_properties.0.disk_address.# = 1
          disk_list.2.device_properties.0.disk_address.0.adapter_type = SCSI
          disk_list.2.device_properties.0.disk_address.0.device_index = 2
          disk_list.2.disk_size_bytes = 209715200
          disk_list.2.disk_size_mib = 200
          disk_list.2.uuid = 704d7286-0d0b-4603-bfc1-869b51b137af
          disk_list.2.volume_group_reference.# = 0
          disk_list.3.data_source_reference.# = 0
          disk_list.3.device_properties.# = 1
          disk_list.3.device_properties.0.device_type = DISK
          disk_list.3.device_properties.0.disk_address.# = 1
          disk_list.3.device_properties.0.disk_address.0.adapter_type = SCSI
          disk_list.3.device_properties.0.disk_address.0.device_index = 3
          disk_list.3.disk_size_bytes = 314572800
          disk_list.3.disk_size_mib = 300
          disk_list.3.uuid = 5db45042-cacf-4d1e-921d-4ec6c715fd22
          disk_list.3.volume_group_reference.# = 0
          enable_script_exec = false
          gpu_list.# = 0
          guest_customization_cloud_init_custom_key_values.% = 0
          guest_customization_cloud_init_meta_data =
          guest_customization_cloud_init_user_data =
          guest_customization_is_overridable = false
          guest_customization_sysprep.% = 0
          guest_customization_sysprep_custom_key_values.% = 0
          guest_os_id =
          hardware_clock_timezone = UTC
          host_reference.% = 2
          host_reference.kind = host
          host_reference.uuid = 8f6b871c-8d31-4f42-873c-1f66079be70e
          memory_size_mib = 186
          metadata.% = 7
          metadata.creation_time = 2018-10-17 20:08:51 +0000 UTC
          metadata.kind = vm
          metadata.last_update_time = 2018-10-17 20:09:02 +0000 UTC
          metadata.name =
          metadata.spec_hash =
          metadata.spec_version = 1
          metadata.uuid = dbc5818c-2e82-44fc-bfd7-2691ed68624d
          name = test-dou-vm-7532894564480512240
          nic_list.# = 0
          num_sockets = 1
          num_vcpus_per_socket = 1
          num_vnuma_nodes = 0
          nutanix_guest_tools.% = 0
          owner_reference.% = 3
          owner_reference.kind = user
          owner_reference.name = admin
          owner_reference.uuid = 00000000-0000-0000-0000-000000000000
          parent_reference.% = 0
          power_state = ON
          power_state_mechanism = HARD
          project_reference.% = 3
          project_reference.kind = project
          project_reference.name = default
          project_reference.uuid = dfae8b2b-7c86-4955-8509-db2d8aa305c8
          should_fail_on_script_failure = false
          state = COMPLETE
          vga_console_enabled = true

          Dependencies:
            data.nutanix_clusters.clusters
            nutanix_image.cirros-034-disk

        Outputs:

        cluster = 27db5fb2-2aa0-48c4-bd2a-2fccdb8fad36
```

Also the `TestAccNutanixVirtualMachine_updateFields` test failed.

Thanks!!

**Comment by simonfuhrer** [comment]
@marinsalinas 

Do you know which operation  requires a poweroff power on?

**Comment by marinsalinas** [comment]
@simonfuhrer update VM requires in some attitudes @JonKohler, Do you know exactly which attributes require the VM turned off to update?.

File: nutanix/resource_nutanix_virtual_machine.go [hunk 1]
```diff
@@ -1585,6 +1590,24 @@ func preFillResUpdateRequest(res *v3.VMResources, response *v3.VMIntentResponse)
 	}
 	res.NicList = nold
 
+	sold := make([]*v3.VMDisk, len(response.Status.Resources.DiskList))
```
Comment (marinsalinas) [inline nutanix/resource_nutanix_virtual_machine.go:1593]: You could keep this and return the fields ForceNew and the rest let them ForceNew as they were before, I think shouldn't delete the DiskList.

**Review by marinsalinas — CHANGES_REQUESTED** [review]

**Comment by JonKohler** [comment]
It would be really interesting to try this again on Prism Central / Prism Element 5.10.0.2 or any of other releases mentioned in field advisory Field Advisory #0059, available here: https://portal.nutanix.com/#/page/static/fieldAdvisories

I have a sneaking suspicion this may be related to that. Might not be, but we should double check

**Pull request closed without merging**